File: backend/src/main/java/com/invoiceautomation/backend/entity/Invoice.java
package com.invoiceautomation.backend.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Lob;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "invoices")
public class Invoice {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column
    private String vendor;

    @Column
    private String invoiceNumber;

    @Column(precision = 12, scale = 2)
    private BigDecimal amount;

    @Column
    private LocalDate invoiceDate;

    @Column(precision = 12, scale = 2)
    private BigDecimal quantity;

    @Column(precision = 12, scale = 2)
    private BigDecimal unitPrice;

    @Column(precision = 12, scale = 2)
    private BigDecimal subtotalAmount;

    @Column(precision = 12, scale = 2)
    private BigDecimal taxAmount;

    @Column(length = 8)
    private String currency;

    @Column
    private LocalDate dueDate;

    @Column(length = 255)
    private String paymentTerms;

    @Lob
    @Column(columnDefinition = "TEXT")
    private String invoiceDescription;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InvoiceStatus status;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InvoiceParseStatus parseStatus;

    @Column(precision = 3, scale = 2)
    private BigDecimal parseConfidence;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InvoiceProcessingStatus processingStatus;

    @Lob
    @Column(columnDefinition = "TEXT")
    private String rawExtractedText;

    @Column(nullable = false)
    private boolean needsReview;

    @Column(nullable = false)
    private boolean duplicateFlag;

    @Column(length = 255)
    private String duplicateReason;

    @Lob
    @Column(columnDefinition = "TEXT")
    private String extractionError;

    @Column(nullable = false)
    private String originalFileName;

    @Column(nullable = false)
    private String storedFileName;

    @Column(nullable = false)
    private String filePath;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    public void onCreate() {
        if (status == null) {
            status = InvoiceStatus.PENDING;
        }
        if (parseStatus == null) {
            parseStatus = InvoiceParseStatus.SUCCESS;
        }
        if (processingStatus == null) {
            processingStatus = InvoiceProcessingStatus.SUCCESS;
        }
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getVendor() {
        return vendor;
    }

    public void setVendor(String vendor) {
        this.vendor = vendor;
    }

    public String getVendorName() {
        return vendor;
    }

    public void setVendorName(String vendorName) {
        this.vendor = vendorName;
    }

    public String getInvoiceNumber() {
        return invoiceNumber;
    }

    public void setInvoiceNumber(String invoiceNumber) {
        this.invoiceNumber = invoiceNumber;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }

    public BigDecimal getTotalAmount() {
        return amount;
    }

    public void setTotalAmount(BigDecimal totalAmount) {
        this.amount = totalAmount;
    }

    public LocalDate getInvoiceDate() {
        return invoiceDate;
    }

    public void setInvoiceDate(LocalDate invoiceDate) {
        this.invoiceDate = invoiceDate;
    }

    public BigDecimal getQuantity() {
        return quantity;
    }

    public void setQuantity(BigDecimal quantity) {
        this.quantity = quantity;
    }

    public BigDecimal getUnitPrice() {
        return unitPrice;
    }

    public void setUnitPrice(BigDecimal unitPrice) {
        this.unitPrice = unitPrice;
    }

    public BigDecimal getSubtotalAmount() {
        return subtotalAmount;
    }

    public void setSubtotalAmount(BigDecimal subtotalAmount) {
        this.subtotalAmount = subtotalAmount;
    }

    public BigDecimal getTaxAmount() {
        return taxAmount;
    }

    public void setTaxAmount(BigDecimal taxAmount) {
        this.taxAmount = taxAmount;
    }

    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public LocalDate getDueDate() {
        return dueDate;
    }

    public void setDueDate(LocalDate dueDate) {
        this.dueDate = dueDate;
    }

    public String getPaymentTerms() {
        return paymentTerms;
    }

    public void setPaymentTerms(String paymentTerms) {
        this.paymentTerms = paymentTerms;
    }

    public String getInvoiceDescription() {
        return invoiceDescription;
    }

    public void setInvoiceDescription(String invoiceDescription) {
        this.invoiceDescription = invoiceDescription;
    }

    public InvoiceStatus getStatus() {
        return status;
    }

    public void setStatus(InvoiceStatus status) {
        this.status = status;
    }

    public InvoiceParseStatus getParseStatus() {
        return parseStatus;
    }

    public void setParseStatus(InvoiceParseStatus parseStatus) {
        this.parseStatus = parseStatus;
    }

    public BigDecimal getParseConfidence() {
        return parseConfidence;
    }

    public void setParseConfidence(BigDecimal parseConfidence) {
        this.parseConfidence = parseConfidence;
    }

    public InvoiceProcessingStatus getProcessingStatus() {
        return processingStatus;
    }

    public void setProcessingStatus(InvoiceProcessingStatus processingStatus) {
        this.processingStatus = processingStatus;
    }

    public String getRawExtractedText() {
        return rawExtractedText;
    }

    public void setRawExtractedText(String rawExtractedText) {
        this.rawExtractedText = rawExtractedText;
    }

    public boolean isNeedsReview() {
        return needsReview;
    }

    public void setNeedsReview(boolean needsReview) {
        this.needsReview = needsReview;
    }

    public boolean isDuplicateFlag() {
        return duplicateFlag;
    }

    public void setDuplicateFlag(boolean duplicateFlag) {
        this.duplicateFlag = duplicateFlag;
    }

    public String getDuplicateReason() {
        return duplicateReason;
    }

    public void setDuplicateReason(String duplicateReason) {
        this.duplicateReason = duplicateReason;
    }

    public String getExtractionError() {
        return extractionError;
    }

    public void setExtractionError(String extractionError) {
        this.extractionError = extractionError;
    }

    public String getOriginalFileName() {
        return originalFileName;
    }

    public void setOriginalFileName(String originalFileName) {
        this.originalFileName = originalFileName;
    }

    public String getStoredFileName() {
        return storedFileName;
    }

    public void setStoredFileName(String storedFileName) {
        this.storedFileName = storedFileName;
    }

    public String getFilePath() {
        return filePath;
    }

    public void setFilePath(String filePath) {
        this.filePath = filePath;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}

File: frontend/src/types/invoice.ts
export type InvoiceStatus = 'PENDING' | 'APPROVED' | 'PAID';
export type InvoiceParseStatus = 'SUCCESS' | 'PARTIAL' | 'FAILED';
export type InvoiceProcessingStatus = 'PROCESSING' | 'SUCCESS' | 'FAILED' | 'DUPLICATE';

export interface Invoice {
  id: number;
  vendor: string | null;
  vendorName: string | null;
  invoiceNumber: string | null;
  amount: number | null;
  totalAmount: number | null;
  quantity: number | null;
  unitPrice: number | null;
  subtotalAmount: number | null;
  taxAmount: number | null;
  currency: string | null;
  invoiceDate: string | null;
  issueDate: string | null;
  dueDate: string | null;
  paymentTerms: string | null;
  invoiceDescription: string | null;
  status: InvoiceStatus;
  parseStatus: InvoiceParseStatus;
  processingStatus: InvoiceProcessingStatus;
  parseConfidence: number | null;
  rawExtractedText: string | null;
  needsReview: boolean;
  duplicateFlag: boolean;
  duplicateReason: string | null;
  extractionError: string | null;
  originalFileName: string;
  createdAt: string;
  updatedAt: string;
}

export interface BatchUploadItemResult {
  fileName: string;
  status: InvoiceProcessingStatus;
  invoiceId: number | null;
  vendorName: string | null;
  invoiceNumber: string | null;
  duplicate: boolean;
  message: string;
}

export interface BatchUploadResult {
  totalFiles: number;
  successCount: number;
  duplicateCount: number;
  failureCount: number;
  results: BatchUploadItemResult[];
  savedInvoices: Invoice[];
}

export interface DashboardSummary {
  totalInvoices: number;
  totalAmount: number;
  pendingCount: number;
  approvedCount: number;
  paidCount: number;
}

export interface InvoiceFilters {
  vendor: string;
  status: '' | InvoiceStatus;
}

export interface InvoiceUploadPayload {
  vendor: string;
  invoiceNumber: string;
  amount: string;
  invoiceDate: string;
  status: InvoiceStatus;
  file: File;
}

export interface InvoiceReviewUpdatePayload {
  invoiceNumber: string | null;
  vendorName: string | null;
  quantity: number | null;
  unitPrice: number | null;
  subtotalAmount: number | null;
  taxAmount: number | null;
  totalAmount: number | null;
  currency: string | null;
  issueDate: string | null;
}

export interface ReviewInvoiceDraft {
  id: number;
  originalFileName: string;
  parseStatus: InvoiceParseStatus;
  parseConfidence: number | null;
  needsReview: boolean;
  rawExtractedText: string | null;
  invoiceNumber: string;
  vendorName: string;
  quantity: string;
  unitPrice: string;
  subtotalAmount: string;
  taxAmount: string;
  totalAmount: string;
  currency: string;
  issueDate: string;
}

export type ReviewInvoiceEditableField =
  | 'invoiceNumber'
  | 'vendorName'
  | 'quantity'
  | 'unitPrice'
  | 'subtotalAmount'
  | 'taxAmount'
  | 'totalAmount'
  | 'currency'
  | 'issueDate';
